### PR TITLE
Add spec for BatchNormGrad

### LINKED ISSUE #345: Add spec for BatchNormGradOp
### Request description

To spec stablehlo.batch_norm_grad

### PR DESCRIPTION
**Pull request by subhankarshah**
Closes #345

### CODE REVIEW

File: docs/spec_draft.md [hunk 1]
```diff
@@ -343,6 +344,78 @@ logical operation.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.batch_norm_grad
+
+### Semantics
+
+Computes gradients of `Batch Normalizing Transform` for `operand`, `scale` and
```
Comment (sdasgup3) [inline docs/spec_draft.md:351]: operand `offset ` does not exists ?

Comment (sdasgup3) [inline docs/spec_draft.md:351]: `Batch Normalizing Transform` is the algorithm referred to Algorithm 1 in the paper. If we want to refer to the same algorithm its better to add the link with the algorithm name. The formula mentioned in the paper seems to be different from the one mentioned in HLO spec. Can you please comment on that?

File: docs/spec_draft.md [hunk 2]
```diff
@@ -343,6 +344,78 @@ logical operation.
...
+offset, for each feature in the `feature_index` dimension, using `grad_output`
```
Comment (sdasgup3) [inline docs/spec_draft.md:352]: Is there a specific reason we mention `grad_output` but not the other operands.

File: docs/spec_draft.md [hunk 3]
```diff
@@ -343,6 +344,78 @@ logical operation.
...
+
+### Outputs
+
+| Name           | Type                                        |
+|----------------|---------------------------------------------|
+| `grad_operand` | tensor of floating-point type               |
+| `grad_scale`   | 1-dimensional tensor of floating-point type |
+| `grad_offset`  | 1-dimensional tensor of floating-point type |
+
+### Constraints
```
Comment (sdasgup3) [inline docs/spec_draft.md:377]: Missing constraints in the spec
https://github.com/openxla/stablehlo/blob/3a89d59082b2a0bb8593931f974256594dadae3a/stablehlo/dialect/StablehloOps.td#L1567

File: stablehlo/dialect/StablehloOps.td [hunk 4]
```diff
@@ -1555,9 +1555,23 @@ def StableHLO_BatchNormGradOp : StableHLO_Op<"batch_norm_grad", [Pure,
...
+    tensors.
 
-    See https://www.tensorflow.org/xla/operation_semantics#batchnormgrad
+    See:
+    https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md#stablehlobatch_norm_grad
+
+    Example:
+    ```mlir
+    %results:3 = "stablehlo.batch_norm_grad"(%operand, %scale, %mean, %variance,
+                                         %grad_output) {
```
Comment (sdasgup3) [inline stablehlo/dialect/StablehloOps.td:1569]: The `%grad_output` is not indented the same way as it is indented in spec.

File: docs/spec_draft.md (hunk 2 above)
Comment (subhankarshah) [reply docs/spec_draft.md:352]: `grad_output` is the gradient that would be "back-propagated", which is why I mentioned it separately.

File: docs/spec_draft.md (hunk 3 above)
Comment (subhankarshah) [reply docs/spec_draft.md:377]: C6, C7

File: docs/spec_draft.md (hunk 1 above)
Comment (subhankarshah) [reply docs/spec_draft.md:351]: grad_offset is computed but offset is not required to compute it or any of `grad_operand` or `grad_scale` for that matter.

Comment (subhankarshah) [reply docs/spec_draft.md:351]: We dont want to refer to that algorithm, but the gradient of that algorithm (the differential equations in the paper).

The formulas are different, but they should be equivalent or the back-prop and gradient descent wont work properly.

**Comment by subhankarshah** [comment]
@sdasgup3 I have resolved the comments with replies or changes, please feel free to un-resolve them.

File: docs/spec_draft.md [hunk 5]
```diff
@@ -343,6 +344,120 @@ logical operation.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.batch_norm_grad
+
+### Semantics
+
+Computes gradients of `operand`, `scale` and offset, using `mean`, `variance`
```
Comment (sdasgup3) [inline docs/spec_draft.md:351]: The `offset` seems out pf place. Add a reference to where it belongs

offset (refer to [stablehlo.batchnorminference](#...))

File: docs/spec_draft.md [hunk 6]
```diff
@@ -343,6 +344,120 @@ logical operation.
...
+| Name           | Type                                        |
+|----------------|---------------------------------------------|
+| `grad_operand` | tensor of floating-point type               |
+| `grad_scale`   | 1-dimensional tensor of floating-point type |
+| `grad_offset`  | 1-dimensional tensor of floating-point type |
+
+### Constraints
+
+  * (C1) 0 $\le$ `feature_index` $\lt$ rank(`operand`).
+  * (C2) `operand`, `scale`, `mean`, `variance`, `grad_output`, `grad_operand`
```
Comment (sdasgup3) [inline docs/spec_draft.md:424]: Drop (C2) as already mentioned in the tables above?

File: docs/spec_draft.md [hunk 7]
```diff
@@ -343,6 +344,120 @@ logical operation.
...
+  (sum,) = reduce(
+      inputs=[operand],
+      init_values=[0],
+      dimensions=[i for i in range(rank(operand)) if i != feature_index],
+      body=lambda x, y: add(x, y))
+  return sum
+
+def compute_mean(operand, feature_index):
+  sum = reduce_sum(operand, feature_index)
+  divisor = constant(num_elements(operand) / dim(operand, feature_index))
```
Comment (sdasgup3) [inline docs/spec_draft.md:368]: drop constant?

File: docs/spec_draft.md [hunk 8]
```diff
@@ -343,6 +344,120 @@ logical operation.
...
+and `grad_output`, the gradient of output from `batch_norm_training`, and
+produces `grad_operand`, `grad_scale` and `grad_offset` tensors. More formally,
+this operation can be expressed as a decomposition to existing StableHLO
+operations using Python-like syntax as follows:
+
+```python
+def reduce_sum(operand, feature_index):
+  (sum,) = reduce(
+      inputs=[operand],
+      init_values=[0],
```
Comment (sdasgup3) [inline docs/spec_draft.md:361]: 0.0

File: docs/spec_draft.md [hunk 9]
```diff
@@ -343,6 +344,120 @@ logical operation.
...
+  scale_bcast = broadcast_in_dim(scale, [feature_index], shape(operand))
+  mean_bcast = broadcast_in_dim(mean, [feature_index], shape(operand))
+  variance_bcast = broadcast_in_dim(variance, [feature_index], shape(operand))
+  epsilon_bcast = broadcast_in_dim(constant(epsilon), [], shape(operand))
+
+  centered_operand = subtract(operand, mean_bcast)
+  stddev = sqrt(add(variance_bcast, epsilon_bcast))
+  normalized_operand = divide(centered_operand, stddev)
+
+  intermediate_cl = compute_mean(multiply(grad_output, divide(normalized_operand, stddev)))
```
Comment (sdasgup3) [inline docs/spec_draft.md:383]: what does the "cl" means? Can we expand the name

File: docs/spec_draft.md [hunk 10]
```diff
@@ -343,6 +344,120 @@ logical operation.
...
+  # Broadcast inputs to shape(operand)
+  scale_bcast = broadcast_in_dim(scale, [feature_index], shape(operand))
+  mean_bcast = broadcast_in_dim(mean, [feature_index], shape(operand))
+  variance_bcast = broadcast_in_dim(variance, [feature_index], shape(operand))
+  epsilon_bcast = broadcast_in_dim(constant(epsilon), [], shape(operand))
+
+  centered_operand = subtract(operand, mean_bcast)
+  stddev = sqrt(add(variance_bcast, epsilon_bcast))
+  normalized_operand = divide(centered_operand, stddev)
+
```
Comment (sdasgup3) [inline docs/spec_draft.md:382]: Add comments here like the code in https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md#stablehlobatch_norm_inference

File: docs/spec_draft.md (hunk 9 above)
Comment (subhankarshah) [reply docs/spec_draft.md:383]: it does not mean anything, cl is the variable name in the XLA semantics, it is just an intermediate variable for simplification purposes.

Comment (subhankarshah) [reply docs/spec_draft.md:383]: I thought of sticking to the name in XLA semantics. Can rename to `intermediate_expr` because we are not linking to the XLA semantics anyway.

File: docs/spec_draft.md (hunk 6 above)
Comment (subhankarshah) [reply docs/spec_draft.md:424]: Tables only say they are of floating type, this is stricter that that, it says they should match exactly.

File: docs/spec_draft.md (hunk 10 above)
Comment (subhankarshah) [reply docs/spec_draft.md:382]: The formula/algorithm here is not intuitive like batch_norm_training, so comments might only state the obvious as to what the code is doing.

File: docs/spec_draft.md (hunk 5 above)
Comment (subhankarshah) [reply docs/spec_draft.md:351]: I will refer to stablehlo.batch_norm_training that would be more accurate.

File: docs/spec_draft.md [hunk 11]
```diff
@@ -343,6 +344,125 @@ logical operation.
...
+//                [[0.1, 0.1], [0.1, 0.1]],
+//                [[0.1, 0.1], [0.1, 0.1]]
+//               ]
+%results:3 = "stablehlo.batch_norm_grad"(%operand, %scale, %mean, %variance,
+                                         %grad_output) {
+  epsilon = 0.0 : f32,
+  feature_index = 2 : i64
+} : (tensor<2x2x2xf32>, tensor<2xf32>, tensor<2xf32>, tensor<2xf32>,
+     tensor<2x2x2xf32>) -> (tensor<2x2x2xf32>, tensor<2xf32>, tensor<2xf32>)
+// %results#0: [
```
Comment (sdasgup3) [inline docs/spec_draft.md:456]: grad_operand  instead of %result#0 and soon?

File: docs/spec_draft.md (hunk 7 above)
Comment (sdasgup3) [reply docs/spec_draft.md:370]: LGTM except for the use of `constant`. The StabbleHLO constant has a implicit shape (provided as output of the op) which is not evident here

File: docs/spec_draft.md [hunk 12]
```diff
@@ -335,6 +336,127 @@ For boolean tensors, computes the logical operation.
...
+    subtract(
+      subtract(grad_output, mean_grad_output_bcast),
+      multiply(intermediate_expr_bcast, centered_operand)
+    )
+  )
+  grad_scale = reduce_sum(
+    multiply(grad_output, normalized_operand), feature_index)
+  grad_offset = reduce_sum(grad_output, feature_index)
+  return grad_operand, grad_scale, grad_offset
+```
```
Comment (burmako) [inline docs/spec_draft.md:396]: Proposed reformulation:

```Python
def compute_sum(operand, feature_index):
  (sum,) = reduce(
      inputs=[operand],
      init_values=[0.0],
      dimensions=[i for i in range(rank(operand)) if i != feature_index],
      body=lambda x, y: add(x, y))
  return sum

def compute_mean(operand, feature_index):
  sum = compute_sum(operand, feature_index)
  divisor = constant(num_elements(operand) / dim(operand, feature_index))
  divisor_bcast = broadcast_in_dim(divisor, [], shape(sum))
  return divide(sum, divisor_bcast)

def batch_norm_grad(operand, scale, mean, variance, grad_output, epsilon, feature_index):
  # Broadcast inputs to shape(operand)
  scale_bcast = broadcast_in_dim(scale, [feature_index], shape(operand))
  mean_bcast = broadcast_in_dim(mean, [feature_index], shape(operand))
  variance_bcast = broadcast_in_dim(variance, [feature_index], shape(operand))
  epsilon_bcast = broadcast_in_dim(constant(epsilon), [], shape(operand))

  # Perform normalization using the provided `mean` and `variance`
  # Intermediate values will be useful for computing gradients
  centered_operand = subtract(operand, mean_bcast)
  stddev = sqrt(add(variance_bcast, epsilon_bcast))
  normalized_operand = divide(centered_operand, stddev)

  # Use the implementation from batchnorm_expander.cc in XLA
  # Temporary variables have exactly the same names as in the C++ code
  elements_per_feature = constant(dim(operand, feature_index))
  i1 = multiply(
    grad_output, 
    broadcast_in_dim(elements_per_feature, [], shape(operand)))
  i2 = broadcast_in_dim(
    compute_sum(grad_output, feature_index), 
    [feature_index], shape(operand))
  i3 = broadcast_in_dim(
    compute_sum(multiply(grad_output, centered_operand)),
    [feature_index], shape(operand))
  i4 = multiply(i3, centered_operand)
  i5 = divide(i4, add(variance_bcast, epsilon_bcast))
  grad_operand = multiply(
    divide(scale_bcast, stddev),
    subtract(subtract(i1, i2), i5))
  )
  grad_scale = compute_sum(
    multiply(grad_output, normalized_operand), feature_index)
  grad_offset = compute_sum(grad_output, feature_index)
  return grad_operand, grad_scale, grad_offset
```

Feedback:
  1. Renamed `reduce_sum` to `compute_sum` to be consistent with other `compute_foo` functions in BatchNorm specs.
  2. Used the implementation for `grad_operand` from batchnorm_expander.cc in XLA, keeping exactly the same temporary variable names: i1, i2, i3, i4 and i5. As you wrote in a nearby comment, these variables don't have good names, so let's align them with something existing for the ease of reviews.

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md [hunk 13]
```diff
@@ -335,6 +336,127 @@ For boolean tensors, computes the logical operation.
...
+
+### Semantics
+
+Computes gradients of `operand`, `scale` and offset
+(refer [batch_norm_training](#stablehlobatch_norm_training)), using `mean`,
+`variance` and `grad_output`, the gradient of output from
+[batch_norm_training](#stablehlobatch_norm_training), and produces
+`grad_operand`, `grad_scale` and `grad_offset` tensors. More formally, this
+operation can be expressed as a decomposition to existing StableHLO operations
+using Python-like syntax as follows:
```
[the remark this answers is not in the document]
Comment (subhankarshah) [reply docs/spec_draft.md:349]: 1. Offset is not an input in this op, which is why backticks were omitted
2. Would it be better to mention that `grad_output` is back propagated or `grad_operand` is computed using it.

File: docs/spec_draft.md (hunk 12 above)
Comment (subhankarshah) [reply docs/spec_draft.md:396]:
```elements_per_feature = constant(
    divide(size(operand), dim(operand, feature_index)))```

Comment (subhankarshah) [reply docs/spec_draft.md:396]:
```python
grad_operand = multiply(
    divide(divide(scale_bcast, stddev), elements_per_feature),
    subtract(subtract(i1, i2), i5))
```

File: docs/spec_draft.md [hunk 14]
```diff
@@ -335,6 +336,127 @@ For boolean tensors, computes the logical operation.
...
+         same shape.
+  * (C5) size(`scale`) $=$ `dim(operand, feature_index)`.
+
+### Examples
+
+```mlir
+// %operand: [
+//            [[1.0, 2.0], [3.0, 4.0]],
+//            [[5.0, 6.0], [7.0, 8.0]]
+//           ]
```
[the remark this answers is not in the document]
Comment (subhankarshah) [reply docs/spec_draft.md:436]: I did that first, but the results were 
```
grad_operand = [[[0, 0], [0, 0]], [[0, 0], [0, 0]]]
grad_scale = [0, 0]
grad_offset = [0.4, 0.4]
```
Which made me feel, I should change it a bit.

File: docs/spec_draft.md (hunk 12 above)
Comment (subhankarshah) [reply docs/spec_draft.md:400]: I had thoroughly checked the previous formulation with `batchnorm_expander` for equivalence.

Comment (subhankarshah) [reply docs/spec_draft.md:400]:
```python
def batch_norm_grad(operand, scale, mean, variance, grad_output, epsilon, feature_index):
  # Broadcast inputs to shape(operand)
  scale_bcast = broadcast_in_dim(scale, [feature_index], shape(operand))
  mean_bcast = broadcast_in_dim(mean, [feature_index], shape(operand))
  variance_bcast = broadcast_in_dim(variance, [feature_index], shape(operand))
  epsilon_bcast = broadcast_in_dim(constant(epsilon), [], shape(operand))
  centered_operand = subtract(operand, mean_bcast)
  stddev = sqrt(add(variance_bcast, epsilon_bcast))
  normalized_operand = divide(centered_operand, stddev)
  intermediate_expr = compute_mean(
    multiply(grad_output, divide(normalized_operand, stddev)), feature_index)
  intermediate_expr_bcast = broadcast_in_dim(
    intermediate_expr, [feature_index], shape(operand))
  mean_grad_output = compute_mean(grad_output, feature_index)
  mean_grad_output_bcast = broadcast_in_dim(
    mean_grad_output, [feature_index], shape(operand))
  grad_operand = multiply(
    divide(scale_bcast, stddev),
    subtract(
      subtract(grad_output, mean_grad_output_bcast),
      multiply(intermediate_expr_bcast, centered_operand)
    )
  )
  grad_scale = compute_sum(
    multiply(grad_output, normalized_operand), feature_index)
  grad_offset = compute_sum(grad_output, feature_index)
  return grad_operand, grad_scale, grad_offset
```

Comment (subhankarshah) [reply docs/spec_draft.md:400]: This is exactly same as outlined in [XLA semantics](https://www.tensorflow.org/xla/operation_semantics#batchnormgrad)

Comment (burmako) [reply docs/spec_draft.md:400]: Thank you for doing the hard work going through the two implementations of BatchNormGrad, making sure they are equivalent and then writing things down!

As we discussed offline, it would be preferable to choose the expander-based version because it's easier to review (since it's running in production, we can be certain that it's correct, while with the version from the Operation Semantics page, figuring that out requires non-trivial work), but I really appreciate the attention to detail.

Agreed with both of your proposed changes! This is something that I missed.

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (subhankarshah) [reply docs/spec_draft.md]: This looks good, thanks.

**Review by burmako — APPROVED** [review]

**Comment by burmako** [comment]
Merging without waiting for CI - the only CI-able change in the PR is documentation within the .td file.
